src/DotNetty.Codecs/MessageToByteEncoder.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs
+{
+    using System;
+    using System.Diagnostics.Contracts;
+    using System.Threading.Tasks;
+    using DotNetty.Buffers;
+    using DotNetty.Common.Utilities;
+    using DotNetty.Transport.Channels;
+
+    public abstract class MessageToByteEncoder<T> : ChannelHandlerAdapter
+    {
+        public virtual bool AcceptOutboundMessage(object message) => message is T;
+
+        public override Task WriteAsync(IChannelHandlerContext context, object message)
+        {
+            Contract.Requires(context != null);
+
+            IByteBuffer buffer = null;
+            Task result;
+            try
+            {
+                if (this.AcceptOutboundMessage(message))
+                {
+                    buffer = this.AllocateBuffer(context);
+                    var input = (T)message;
+                    try
+                    {
+                        this.Encode(context, input, buffer);
+                    }
+                    finally
+                    {
+                        ReferenceCountUtil.Release(input);
+                    }
+
+                    if (buffer.IsReadable())
+                    {
+                        result = context.WriteAsync(buffer);
+                    }
+                    else
+                    {
+                        buffer.Release();
+                        result = context.WriteAsync(Unpooled.Empty);
+                    }
+
+                    buffer = null;
+                }
+                else
+                {
+                    return context.WriteAsync(message);
+                }
+            }
+            catch (EncoderException e)
+            {
+                return TaskEx.FromException(e);
+            }
+            catch (Exception ex)
+            {
+                return TaskEx.FromException(new EncoderException(ex)); // todo: we don't have a stack on EncoderException but it's present on inner exception.
```

Comment:
I will remove it. I was wondering something like unwrap task aggregated exception.